src/ui/sign_message_nbgl.c

```diff
@@ -0,0 +1,145 @@
+#ifdef HAVE_NBGL
+
+#include "io.h"
+#include "menu.h"
+#include "sol/parser.h"
+#include "sol/printer.h"
+#include "sol/print_config.h"
+#include "sol/message.h"
+#include "sol/transaction_summary.h"
+#include "glyphs.h"
+#include "apdu.h"
+#include "utils.h"
+
+#include "nbgl_page.h"
+#include "nbgl_use_case.h"
+
+#include "handle_sign_message.h"
+#include "handle_sign_offchain_message.h"
+
+// Layout of the review flow
+static nbgl_layoutTagValueList_t layout;
+// Used by NBGL to display the reference the pair number N
+static nbgl_layoutTagValue_t current_pair;
+
+// We will display at most 4 items on a Stax review screen
+#define MAX_SIMULTANEOUS_DISPLAYED_SLOTS 4
+typedef struct dynamic_slot_s {
+    char title[sizeof(G_transaction_summary_title)];
+    char text[sizeof(G_transaction_summary_text)];
+} dynamic_slot_t;
+static dynamic_slot_t displayed_slots[MAX_SIMULTANEOUS_DISPLAYED_SLOTS];
+
+// Final review screen of the message signing flow
+static nbgl_pageInfoLongPress_t review_final_long_press;
+
+// Callback called when the user confirms the message rejection
+static void rejectChoice(void) {
+    sendResponse(0, false, false);
+    nbgl_useCaseStatus("message\nrejected", false, ui_idle);
+}
+
+// If the user asks for message rejection, ask for confirmation
+static void rejectUseCaseChoice(void) {
+    nbgl_useCaseConfirm("Reject message?", NULL, "Yes, reject", "Go back to message", rejectChoice);
+}
+
+// Callback called on the final review screen
+static void review_final_callback(bool confirmed) {
+    if (confirmed) {
+        sendResponse(set_result_sign_message(), true, false);
+        nbgl_useCaseStatus("MESSAGE\nSIGNED", true, ui_idle);
+    } else {
+        rejectUseCaseChoice();
+    }
+}
+
+// NBGL library has to know how many steps will be displayed
+static size_t transaction_steps_number;
+static bool last_step_is_ascii;
+
+// function called by NBGL to get the current_pair indexed by "index"
+// current_pair will point at values stored in displayed_slots[]
+// this will enable displaying at most sizeof(displayed_slots) values simultaneously
+static nbgl_layoutTagValue_t *get_single_action_review_pair(uint8_t index) {
+    dynamic_slot_t *slot = &displayed_slots[index % ARRAY_COUNT(displayed_slots)];
```

Comment:
Why do you need a pointer here ?

Reply:
We want to modify the static slot so the data survives this functions return for the UI